src/execution/command_parser.c

```diff
@@ -1,98 +1,129 @@
+                                          RedisModuleString ***inkeys, RedisModuleString ***outkeys,
+                                          RedisModuleString **runkey, long long *timeout) {
 
-    if (argc < 6) {
+    if (argc < 8) {
         RAI_SetError(error, RAI_EMODELRUN,
-                     "ERR wrong number of arguments for 'AI.MODELRUN' command");
+                     "ERR wrong number of arguments for 'AI.MODELEXECUTE' command");
         return REDISMODULE_ERR;
     }
-    size_t argpos = 1;
-    const int status = RAI_GetModelFromKeyspace(ctx, argv[argpos], model, REDISMODULE_READ, error);
+    size_t arg_pos = 1;
+    const int status = RAI_GetModelFromKeyspace(ctx, argv[arg_pos], model, REDISMODULE_READ, error);
     if (status == REDISMODULE_ERR) {
         return REDISMODULE_ERR;
     }
-    RAI_HoldString(NULL, argv[argpos]);
-    *runkey = argv[argpos];
-    const char *arg_string = RedisModule_StringPtrLen(argv[++argpos], NULL);
+    *runkey = RAI_HoldString(NULL, argv[arg_pos++]);
+    const char *arg_string = RedisModule_StringPtrLen(argv[arg_pos++], NULL);
 
-    // Parse timeout arg if given and store it in timeout
-    if (!strcasecmp(arg_string, "TIMEOUT")) {
-        if (_parseTimeout(argv[++argpos], error, timeout) == REDISMODULE_ERR)
-            return REDISMODULE_ERR;
-        arg_string = RedisModule_StringPtrLen(argv[++argpos], NULL);
-    }
     if (strcasecmp(arg_string, "INPUTS") != 0) {
         RAI_SetError(error, RAI_EMODELRUN, "ERR INPUTS not specified");
         return REDISMODULE_ERR;
     }
 
-    bool is_input = true, is_output = false;
-    size_t ninputs = 0, noutputs = 0;
-
-    while (++argpos < argc) {
-        arg_string = RedisModule_StringPtrLen(argv[argpos], NULL);
-        if (!strcasecmp(arg_string, "OUTPUTS") && !is_output) {
-            is_input = false;
-            is_output = true;
-        } else {
-            RAI_HoldString(NULL, argv[argpos]);
-            if (is_input) {
-                ninputs++;
-                *inkeys = array_append(*inkeys, argv[argpos]);
-            } else {
-                noutputs++;
-                *outkeys = array_append(*outkeys, argv[argpos]);
-            }
-        }
+    long long ninputs = 0, noutputs = 0;
+    if (RedisModule_StringToLongLong(argv[arg_pos++], &ninputs) != REDISMODULE_OK) {
+        RAI_SetError(error, RAI_EMODELRUN, "ERR Invalid argument for input_count");
+        return REDISMODULE_ERR;
+    }
+    if (ninputs <= 0) {
+        RAI_SetError(error, RAI_EMODELRUN, "ERR Input count must be a positive integer");
+        return REDISMODULE_ERR;
     }
     if ((*model)->ninputs != ninputs) {
         RAI_SetError(error, RAI_EMODELRUN,
                      "Number of keys given as INPUTS here does not match model definition");
         return REDISMODULE_ERR;
     }
+    // arg_pos = 4 at this point, as we always have 4 args before the input keys.
+    for (; arg_pos < ninputs + 4 && arg_pos < argc; arg_pos++) {
```

Comment:
have a dedicated condition for `argpos + inputs >= argc` and return an error in this case. remove it from the loop
also, have a local variable `size_t inputs_start = 4` and in the loop check `argpos < input_start + ninputs` on something similar.